### PR TITLE
Support launching Cypress tests in Podman on Ubuntu

### PR DESCRIPTION
**Pull request by andybalaam**
This is the recipe I found I needed to get the tests to run on Ubuntu using Podman instead of Docker.

Would be interested in feedback from @weeman1337, who added the initial Podman support, which did not work for me. (But he is away this week, so I don't suggest we wait for him to do the review.)

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

### CODE REVIEW

File: cypress/plugins/docker/index.ts [hunk 1]
```diff
@@ -36,12 +36,16 @@ export async function dockerRun(opts: {
...
-        params.push("-u", `${userInfo.uid}:${userInfo.gid}`);
-
         if (await isPodman()) {
-            // keep the user ID if the docker command is actually podman
-            params.push("--userns=keep-id");
+            // In podman, run as root in the container, so we're the current
+            // user on the host
+            params.push("-u", "0:0");
+            params.push("-e", "UID=0");
+            params.push("-e", "GID=0");
```
Comment (richvdh) [inline cypress/plugins/docker/index.ts:44]:
```suggestion
            # Tell the Synapse container not to switch UID
            params.push("-e", "UID=0");
            params.push("-e", "GID=0");
```

File: cypress/plugins/docker/index.ts [hunk 2]
```diff
@@ -36,12 +36,16 @@ export async function dockerRun(opts: {
...
     if (params?.includes("-v") && userInfo.uid >= 0) {
-        // On *nix we run the docker container as our uid:gid otherwise cleaning it up its media_store can be difficult
-        params.push("-u", `${userInfo.uid}:${userInfo.gid}`);
-
         if (await isPodman()) {
-            // keep the user ID if the docker command is actually podman
-            params.push("--userns=keep-id");
+            // In podman, run as root in the container, so we're the current
+            // user on the host
+            params.push("-u", "0:0");
```
Comment (richvdh) [inline cypress/plugins/docker/index.ts:42]: is this not the default? https://docs.podman.io/en/latest/markdown/podman-run.1.html#user-u-user-group seems to say so.

(I don't mind if you want to make it explicit, but please comment that is what you are doing)

File: cypress/plugins/docker/index.ts [hunk 3]
```diff
@@ -36,12 +36,16 @@ export async function dockerRun(opts: {
...
 
     if (params?.includes("-v") && userInfo.uid >= 0) {
-        // On *nix we run the docker container as our uid:gid otherwise cleaning it up its media_store can be difficult
-        params.push("-u", `${userInfo.uid}:${userInfo.gid}`);
-
         if (await isPodman()) {
-            // keep the user ID if the docker command is actually podman
-            params.push("--userns=keep-id");
+            // In podman, run as root in the container, so we're the current
+            // user on the host
```
Comment (richvdh) [inline cypress/plugins/docker/index.ts:41]:
```suggestion
            // In podman, run as root in the container, which maps to the current
            // user on the host
```

I'm not a podman user, but my understanding that this is only true in some deployments? https://docs.podman.io/en/latest/markdown/podman-run.1.html#user-u-user-group says it only applies "when a user namespace is in use" - do we need to arrange for that to be the case?

File: cypress/plugins/docker/index.ts (hunk 2 above)
Comment (andybalaam) [reply cypress/plugins/docker/index.ts:42]: I'm not sure. I don't read that page as "root is the default".

File: cypress/plugins/docker/index.ts (hunk 3 above)
Comment (andybalaam) [reply cypress/plugins/docker/index.ts:41]: Yeah you're right. This is for rootless podman only, where user namespaces are always in use. I'll add a comment.

File: cypress/plugins/docker/index.ts (hunk 2 above)
Comment (richvdh) [reply cypress/plugins/docker/index.ts:42]: > Without this argument, the command will run as the user specified in the container image. Unless overridden by a USER command in the Containerfile or by a value passed to this option, this user generally defaults to root.

Synapse doesn't set `USER`, so it sounds like it defaults to root to me?

(Though I wish the 'generally' was a bit more confident)

Comment (andybalaam) [reply cypress/plugins/docker/index.ts:44]: Ah, I see what you mean. I'll add a comment.

File: cypress/plugins/docker/index.ts [hunk 4]
```diff
@@ -36,12 +36,22 @@ export async function dockerRun(opts: {
...
     if (params?.includes("-v") && userInfo.uid >= 0) {
-        // On *nix we run the docker container as our uid:gid otherwise cleaning it up its media_store can be difficult
-        params.push("-u", `${userInfo.uid}:${userInfo.gid}`);
-
         if (await isPodman()) {
-            // keep the user ID if the docker command is actually podman
-            params.push("--userns=keep-id");
+            // Note: this setup is for podman rootless containers.
+
+            // In podman, run as root in the container, so we're the current
```
Comment (richvdh) [inline cypress/plugins/docker/index.ts:42]: I think this is clearer:

```suggestion
            // In podman, run as root in the container, which maps to the current
```

File: cypress/plugins/docker/index.ts [hunk 5]
```diff
@@ -36,12 +36,22 @@ export async function dockerRun(opts: {
...
+            // Dockerfile doesn't specify, but we're being explicit here
+            // because it's important for the permissions to work.
+            params.push("-u", "0:0");
+
+            // Tell Synapse not to switch UID
+            params.push("-e", "UID=0");
+            params.push("-e", "GID=0");
+        } else {
+            // On *nix we run the docker container as our uid:gid - otherwise,
+            // cleaning it up its media_store can be difficult
```
Comment (richvdh) [inline cypress/plugins/docker/index.ts:53]: I think this comment is relevant to both podman and docker, though the means of achieving it is different between the two.

Suggest moving out of the `if/else` condition.

**Review by richvdh — APPROVED** [review]
ltgm otherwwise

File: cypress/plugins/docker/index.ts (hunk 5 above)
Comment (andybalaam) [reply cypress/plugins/docker/index.ts]: Done

**Comment by weeman1337** [comment]
Works for me :+1:

**Pull request merged**